jitsi-media-transform/src/main/kotlin/org/jitsi/nlj/rtp/bandwidthestimation2/GoogCcTransportCcEngine.kt

```diff
@@ -0,0 +1,359 @@
+            }
+            maybeScheduleProbing(now)
+        }
+    }
+
+    @Synchronized
+    override fun mediaPacketSent(tccSeqNum: Int, length: DataSize) {
+        val now = clock.instant()
+        // We need to do sequence number unwrapping in the TransportFeedbackAdapter, so
+        // truncate it here so we can unwrap it again later.
+        val truncatedSeqNum = tccSeqNum and 0xFFFF
+        val sentPacketInfo = SentPacketInfo(
+            packetId = truncatedSeqNum,
+            sendTime = now,
+            info = PacketInfo(
+                // TODO I think these should always be true when tccSeqNum is defined?
+                includedInAllocation = true,
+                includedInFeedback = true,
+                packetSizeBytes = length.bytes.toLong()
+            )
+        )
+        val sentPacket = feedbackAdapter.processSentPacket(sentPacketInfo)
+        if (sentPacket != null) {
+            val update = networkController.onSentPacket(sentPacket)
+            processUpdate(update)
+        }
+    }
+
+    @Synchronized
+    override fun getStatistics(): StatisticsSnapshot {
+        val now = clock.instant()
+        return StatisticsSnapshot(
+            feedbackAdapter.getStatisitics(),
+            (networkController as GoogCcNetworkController).getStatistics(now)
+        )
+    }
+
+    @Synchronized
+    override fun addBandwidthListener(listener: BandwidthListener) {
+        listeners.add(listener)
+    }
+
+    @Synchronized
+    override fun removeBandwidthListener(listener: BandwidthListener) {
+        listeners.remove(listener)
+    }
+
+    override fun start() {
+        val startTime = clock.instant()
+        var update =
+            networkController.onNetworkAvailability(NetworkAvailability(atTime = startTime, networkAvailable = true))
+        processUpdate(update) // Does this make sense to do during init?
+
+        update = networkController.onProcessInterval(ProcessInterval(atTime = startTime))
+        processUpdate(update)
+
+        val processInterval = factory.getProcessInterval().toMillis()
+
+        processTask = scheduledExecutor.scheduleAtFixedRate({
+            synchronized(this@GoogCcTransportCcEngine) {
+                val now = clock.instant()
+                update = networkController.onProcessInterval(ProcessInterval(atTime = now))
+                processUpdate(update)
+            }
+        }, processInterval, processInterval, TimeUnit.MILLISECONDS)
+    }
+
+    override fun stop() {
+        synchronized(this@GoogCcTransportCcEngine) {
+            // Stop bitrateProber from initiating any new probes
+            bitrateProber.setEnabled(false)
+            probeTask?.cancel(false)
+        }
+        processTask?.cancel(false)
+    }
+
+    private fun processUpdate(update: NetworkControlUpdate) {
+        update.targetRate?.let { targetRate ->
+            logger.debug("GoogleCcEstimator setting TargetRate to $targetRate")
```

Comment:
Suggest using `logger.debug { }` to avoid string interpolation when debug is disabled